File: api-registro-escolar/src/main/java/com/dt/registroescolar/api_registro_escolar/service/EstudianteService.java
package com.dt.registroescolar.api_registro_escolar.service;

import com.dt.registroescolar.api_registro_escolar.dto.EstudianteDTO;

import java.util.List;
import java.util.Optional;

public interface EstudianteService {

    EstudianteDTO registrarEstudiante(EstudianteDTO estudianteDTO);

    List<EstudianteDTO> listarEstudiantes();

    Optional<EstudianteDTO> buscarPorId(Long id);

    Optional<EstudianteDTO> buscarPorNumeroMatricula(String numeroMatricula);

    EstudianteDTO actualizarEstudiante(Long id, EstudianteDTO estudianteDTO);

    void eliminarEstudiante(Long id);
}

File: registro-escolar-frontend/src/app/services/estudiante.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { EstudianteDTO } from '../models/estudiante.dto';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class EstudianteService {
  private baseUrl = 'http://localhost:8080/api/estudiantes';

  constructor(private http: HttpClient) {}

  registrarEstudiante(estudiante: EstudianteDTO): Observable<EstudianteDTO> {
    return this.http.post<EstudianteDTO>(`${this.baseUrl}/registrar`, estudiante);
  }

  listarEstudiantes(): Observable<EstudianteDTO[]> {
    return this.http.get<EstudianteDTO[]>(this.baseUrl);
  }

  obtenerEstudiantePorId(id: number): Observable<EstudianteDTO> {
    return this.http.get<EstudianteDTO>(`${this.baseUrl}/buscar/id/${id}`);
  }

  buscarPorGrado(grado: string): Observable<EstudianteDTO[]> {
    return this.http.get<EstudianteDTO[]>(`${this.baseUrl}/buscar/grado/${grado}`);
  }

  actualizarEstudiante(estudiante: EstudianteDTO): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/actualizar/${estudiante.id}`, estudiante);
  }

  eliminarEstudiante(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/eliminar/${id}`);
  }
}
